File: python/day4.py
from utils import file_readlines
import re

input_file = open('../_inputs/day4.txt', 'r')
lines = [line.strip().replace("\n", " ") for line in input_file.read().split('\n\n')]

REQUIRED_FIELDS = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
BIRTH_YEAR_ALLOWED_RANGE = range(1920, 2003)
ISSUE_YEAR_ALLOWED_RANGE = range(2010, 2021)
EXPIRATION_YEAR_ALLOWED_RANGE = range(2020, 2031)
HEIGHT_REGEX = "^(?P<value>\d+)(?P<unit>cm|in)$"
HAIR_COLOR_REGEX = "^#([0-9a-f]{6})$"
PASSPORT_ID_REGEX = "^\d{9}$"
EYE_COLORS = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]


def run_part1():
    passports = (line.split(' ') for line in lines)

    passport_fields = lambda passport: (field.split(":")[0] for field in passport)
    passport_has_field = lambda passport, field: field in passport_fields(passport)

    valid = filter(lambda passport: all(passport_has_field(passport, field) for field in REQUIRED_FIELDS), passports)

    print(sum(1 for _ in valid))


def run_part2():
    passports = map(lambda line: parse_passport(line), [line.split(' ') for line in lines])

    valid = filter(lambda passport: is_passport_valid(passport), passports)

    print(sum(1 for _ in valid))


def parse_passport(line):
    fields = (field.split(':') for field in line)

    return {key : value for key, value in fields}


def is_passport_valid(passport):
    return all(is_field_valid(passport, field) for field in REQUIRED_FIELDS)


def is_field_valid(passport, field):
    validations = {
        'byr': lambda value: value.isnumeric() and int(value) in BIRTH_YEAR_ALLOWED_RANGE,
        'iyr': lambda value: value.isnumeric() and int(value) in ISSUE_YEAR_ALLOWED_RANGE,
        'eyr': lambda value: value.isnumeric() and int(value) in EXPIRATION_YEAR_ALLOWED_RANGE,
        'hgt': lambda value: is_valid_height(value),
        'hcl': lambda value: re.match(HAIR_COLOR_REGEX, value),
        'ecl': lambda value: value in EYE_COLORS,
        'pid': lambda value: re.match(PASSPORT_ID_REGEX, value),
        'cid': lambda value: True
    }

    return field in passport and validations[field](passport[field])


def is_valid_height(height):
    captures = re.findall(HEIGHT_REGEX, height)

    validations = {
        'cm': lambda value: int(value) in range(150, 194),
        'in': lambda value: int(value) in range(59, 77)
    }

    if len(captures) == 1:
        value, unit = captures[0]

        return validations[unit](value)

    return False

File: rust/src/day4.rs
use regex::Regex;
use std::ops::RangeInclusive;

use super::utils;

const REQUIRED_FIELDS: [&str; 7] = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"];

const BIRTH_YEAR_ALLOWED_RANGE: RangeInclusive<i16> = 1920..=2002;
const ISSUE_YEAR_ALLOWED_RANGE: RangeInclusive<i16> = 2010..=2020;
const EXPIRATION_YEAR_ALLOWED_RANGE: RangeInclusive<i16> = 2020..=2030;
const HEIGHT_REGEX: &str = r"^(?P<value>\d+)(?P<unit>cm|in)$";
const HAIR_COLOR_REGEX: &str = r"^#([0-9a-f]{6})$";
const PASSPORT_ID_REGEX: &str = r"^\d{9}$";
const EYE_COLORS: [&str; 7] = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"];

pub fn run_part1() -> usize {
  let result = utils::file::read_to_string("../_inputs/day4.txt")
    .trim()
    .split("\n\n")
    .map(|line| line.replace("\n", " "))
    .filter(|line| {
      let passport_fields: Vec<&str> = line
        .split(" ")
        .map(|field| field.split(":").collect::<Vec<&str>>()[0])
        .collect();

      REQUIRED_FIELDS
        .iter()
        .all(|required_field| passport_fields.contains(required_field))
    })
    .count();

  println!("{}", result);

  result
}

pub fn run_part2() -> usize {
  let result = utils::file::read_to_string("../_inputs/day4.txt")
    .trim()
    .split("\n\n")
    .map(|line| line.replace("\n", " "))
    .filter(|line| {
      let passport_fields: Vec<Vec<&str>> = line
        .split(" ")
        .map(|field| field.split(":").collect::<Vec<&str>>())
        .collect();

      REQUIRED_FIELDS.iter().all(|required_field| {
        match passport_fields
          .iter()
          .find(|field| &field[0] == required_field)
        {
          None => false,
          Some(field) => is_field_valid(field[0], field[1]),
        }
      })
    })
    .count();

  println!("{}", result);

  result
}

fn is_field_valid(key: &str, val: &str) -> bool {
  match key {
    "byr" => is_integer_in_range(val, BIRTH_YEAR_ALLOWED_RANGE),
    "iyr" => is_integer_in_range(val, ISSUE_YEAR_ALLOWED_RANGE),
    "eyr" => is_integer_in_range(val, EXPIRATION_YEAR_ALLOWED_RANGE),
    "hgt" => is_valid_height(val),
    "hcl" => is_match(val, HAIR_COLOR_REGEX),
    "ecl" => EYE_COLORS.contains(&val),
    "pid" => is_match(val, PASSPORT_ID_REGEX),
    "cid" => true,
    _ => false,
  }
}

fn is_integer_in_range(val: &str, range: RangeInclusive<i16>) -> bool {
  match val.parse::<i16>() {
    Ok(num) => range.contains(&num),
    _ => false,
  }
}

fn is_valid_height(val: &str) -> bool {
  match Regex::new(HEIGHT_REGEX).unwrap().captures(val) {
    None => false,
    Some(captures) => {
      let value = captures.name("value").unwrap().as_str();
      let unit = captures.name("unit").unwrap().as_str();

      match unit {
        "cm" => is_integer_in_range(value, 150..=193),
        "in" => is_integer_in_range(value, 59..=76),
        _ => false,
      }
    }
  }
}

fn is_match(val: &str, regex: &str) -> bool {
  Regex::new(regex).unwrap().is_match(val)
}
